tools/git-change-exec/main.go

```diff
@@ -0,0 +1,307 @@
+
+			for _, cat := range args {
+				if len(actions[cat]) == 0 {
+					log.Fatalf("could not find actions for %s", cat)
+				}
+
+				gce.actionsToCheck = append(gce.actionsToCheck, actions[cat]...)
+			}
+
+			currentPath, err := os.Getwd()
+			if err != nil {
+				log.Fatalf("getting current working directory: %v", err)
+			}
+
+			defer func() {
+				err := os.Chdir(currentPath)
+				if err != nil {
+					log.Printf("could not change back to previous dir %s: %v", currentPath, err)
+				}
+			}()
+
+			gce.g, err = git.PlainOpenWithOptions("./", &git.PlainOpenOptions{DetectDotGit: true})
+			if err != nil {
+				log.Fatalf("open git path %s failed: %v", gce.gitPath, err)
+			}
+
+			wt, err := gce.g.Worktree()
+			if err != nil {
+				log.Fatalf("could not determine worktree: %v", err)
+			}
+			gce.gitPath = wt.Filesystem.Root()
+
+			err = os.Chdir(gce.gitPath)
+			if err != nil {
+				log.Fatalf("could not change to %s: %v", gce.gitPath, err)
+			}
+
+			gce.fetchOrigin()
+
+			gce.collectActionsGitTree()
+			gce.collectDirtyGitTree()
+
+			gce.runActionDos()
+		},
+	}
+
+	rootCmd.PersistentFlags().BoolVarP(&dryRun, "dry-run", "d", false, "")
+	err := rootCmd.Execute()
+	if err != nil {
+		log.Fatalf("corba failed with: %v", err)
+	}
+
+}
+
+func (gce *gitChangeExec) fetchOrigin() {
+	err := gce.g.Fetch(&git.FetchOptions{
+		RemoteName: "origin",
+		Tags:       git.AllTags,
+	})
+	if err != nil {
+		debugLog("fetching from origin failed: %v", err)
+	}
+}
+
+func (gce *gitChangeExec) collectActionsGitTree() {
+
+	logIter, err := gce.g.Log(&git.LogOptions{})
+	if err != nil {
+		log.Fatalf("getting log failed: %v", err)
+	}
+
+	branchHead, err := logIter.Next()
+	if err != nil {
+		log.Fatalf("getting log.Next failed: %v", err)
+	}
+
+	commonBase := gce.findCommonBase(branchHead)
+
+	logIter, err = gce.g.Log(&git.LogOptions{})
+	if err != nil {
+		log.Fatalf("getting log for iteration failed: %v", err)
+	}
+
+	err = logIter.ForEach(func(c *object.Commit) error {
+		for _, cb := range commonBase {
+			if c.Hash == cb.Hash {
+				return storer.ErrStop
+			}
+		}
+
+		commitStats, err := c.Stats()
+		if err != nil {
+			log.Fatalf("getting commit stats failed: %v", err)
+		}
+
+		for _, st := range commitStats {
+			gce.addActionByPath(st.Name)
+		}
+
+		return nil
+	})
+	logIter.Close()
+
+	if err != nil {
+		log.Fatalf("iterating over commits failed: %v", err)
+	}
+}
+
+func (gce *gitChangeExec) findCommonBase(branchHead *object.Commit) []*object.Commit {
+	var commonBase []*object.Commit
+	masterRef := gce.retrieveMasterRef()
+
+	refs := masterRef
+	refs = append(refs, gce.retrieveLtsRefs()...)
+
+	for _, ref := range refs {
+		commit, err := gce.g.CommitObject(ref.Hash())
+		if err != nil {
+			log.Printf("retrieve commit object from ref %v failed: %v", ref, err)
+			continue
+		}
+		commonBase = append(commonBase, commit)
+		addBase, err := branchHead.MergeBase(commit)
+		if err != nil {
+			log.Printf("finding merge base failed: %v", err)
```

Comment:
I guess this will fail in most cases as you will try to find a merge base for all the potential target branches here. If the branch is based on `master` and you try to find a merge base for `-stable` - the result is kind of undefined and irrelevant... Maybe we can still call it here, but the error printing is unnecessary, I guess.

Reply:
I removed the error printing, now it only does `debugLog`